### PR TITLE
Feature/1825 diff filtering

### PR DESCRIPTION
**Pull request by Nochdarus**
Hello!
Sometime ago I got disappointed in SourceTree (instability, closed sources).
And I decided to move to this nice tool of yours. I liked overall relatively high stability and certain features.

But I have to compare different states very often and interested only in certain file types (but not just exact files), that's why nice Search option in diff wasn't enough. 

So I implemented the filter as well like was suggested here https://github.com/gitextensions/gitextensions/issues/1825. 
Also while at it I slightly separated update of ListView and dictionary behind it. There was also a very minor bug with "No changes"  label not showing in case if states are the same. I corrected that.

Here are some screenshots (as you requested in CONTRIBUTING.md):
diff_filter
filter_applied
regex_error_message

I would be happy if you did a review and merged my changes.

P.S. Also in another branch I did minor visibility adjustments, which you might find okay (or may be not). I'll make it a separate merge request.

P.P.S. Eventually I would like to adjust the graph in general and add some indication for loading diff, but that's another topic.

Anyhow, thanks for a nice product!

### CODE REVIEW

**Comment by KindDragon** [comment]
I am not against filtering, but I do not know about filtering box. Filtering is a rare problem for most users.
I think maybe show it by hotkey?

**Comment by Nochdarus** [comment]
@KindDragon  I think it depends if we (and/or majority of users) want other features for diff window (like a tree display or sorting by path). Those are not a problem to implement.

My point is that then we can aggregate those things with the filtering. How - another question.

If we don't implement those, then yes, shortcut makes sense (or context menu, for the search you did both). Or checkbox-ish entry in context menu probably would be OK, like Show filter box
showfilterbox

The only not nice thing about it is that we have options applying both to one file and to the whole list in one menu (i.e. now it is only Find option and now Show filter box). But I think it is minor.

**Comment by Radrik5** [comment]
Maybe we can completely replace Find with Filter (move focus to the filter box on Ctrl+F). Filter enables you to do the same as Find plus a bit more. And the visible filter box is much more intuitive then the Find edit box opened on Ctrl+F or from the context menu.

**Comment by Radrik5** [comment]
@Nochdarus could you please add cue banner "Filter files..." to the filter box to make it more intuitive?

**Comment by Nochdarus** [comment]
@Radrik5  at the moment instead of cue banner there is just a magnifying glass picture which works like cue banner in terms of appearing and disappearing.

I though about adding "Filter files..." cue banner or something similar, but those reluctant to do it because of translation. If you also feel that it makes sense to have it like this (anyway, developers should know English) then yeah, no problem.

**Comment by vbjay** [comment]
Don't worry about translation.  Others will translate it.

**Comment by Nochdarus** [comment]
Updated it to "Filter files..." with LightGray color:
filter files cue banner

So, what about showing it? 

Ctrl+F and focus like @Radrik5 proposed?
Or CheckBox in context menu?
Or something else?

Another point: what should happen to file diff browsing window on the right? At the moment I don't update it is contents if a filtering action is performed. This way it is quicker.

**Comment by jbialobr** [comment]
> Ctrl+F and focus like @Radrik5 proposed?

I wouldn't remove the existing Find dialog as it provides a different functionality than the filter box and I like them both. 

>Or CheckBox in context menu?

I like this option, plus some hotkey eg. ctrl+shift+f to show/hide the filter input (hide automatically should clear the filter)

>Another point: what should happen to file diff browsing window on the right? At the moment I don't update it is contents if a filtering action is performed. This way it is quicker.

Nothing special. When a file is selected and satisfies the filter it should remain selected so there is no need to refresh the diff view. When a file is selected and it stops satisfying the filter the selection should be cleared, so do the diff view.

**Comment by jbialobr** [comment]
Can you wonder if it is doable to incorporate the filter box to the FileStatusList. I would be happy to have the filter box in the Commit window.

**Comment by EbenZhang** [comment]
Hi @Nochdarus, good work.

Is it possible to make the filter case insensative?

**Comment by Nochdarus** [comment]
Following updates are made:
1) Rebase on the master of the original repo
2) Case insensitive diff file filter is merged (thanks for the work to @EbenZhang )
3) As @jbialobr suggested, selected item is preserved in case if it still satisfies the filtering (also I disabled multiselection to make preservation simpler, we have no use for multiselection anyway)
4) Also as @jbialobr suggested, filtering is now inside FileStatusList, thus we get filters in commit window as well.
However at the moment filter has no effect on Stage/Unstage All (i.e. all files will be staged/unstaged regardless of filtering). Do you think it is okay? (Or it can be helpful to stage for example only certain file extensions)

Side note 1:
AppVeyor fails, but it looks like a problem on the server (read Details for that https://ci.appveyor.com/project/jbialobr/gitextensions/build/2.49.13)

Side note 2:
Also, @jbialobr, @EbenZhang  I would really appreciate your opinion here https://github.com/gitextensions/gitextensions/issues/3203#issuecomment-238206745

**Comment by jbialobr** [comment]
Nice, but there is somewhere a performance issue. Do a stash that includes untracked files in the GitExt repo and try to see the diff for the stash merge (this diff shows all the project files). It takes about 6 seconds to see unfiltered files. The 2.48 version shows these same files in less than a second.

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -72,8 +72,11 @@ public FileStatusList()
             FileStatusListView.SmallImageList = _images;
             FileStatusListView.LargeImageList = _images;
 
-            NoFiles.Visible = false;
+            HandleVisibility_NoFilesLabel_FilterComboBox(filesPresent: true);
+            this.Controls.SetChildIndex(NoFiles, 0);
             NoFiles.Font = new Font(SystemFonts.MessageBoxFont, FontStyle.Italic);
+
+            _filter = new Regex(".*");
```
Comment (jbialobr) [inline GitUI/UserControls/FileStatusList.cs]: I would assign a null value when the filter value is empty. The condition for accepting files should accept all files when the `_filter` is null.

File: GitUI/UserControls/FileStatusList.cs [hunk 2]
```diff
@@ -332,7 +335,7 @@ public IEnumerable<GitItemStatus> SelectedItems
 
         [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
         [Browsable(false)]
-        public GitItemStatus SelectedItem
+       public GitItemStatus SelectedItem
```
Comment (jbialobr) [inline GitUI/UserControls/FileStatusList.cs]: One space is missing here.

File: GitUI/UserControls/FileStatusList.cs [hunk 3]
```diff
@@ -781,6 +822,15 @@ public void SetDiff(GitRevision revision)
                 else
                 {
                     GitItemsWithParents dictionary = new Dictionary<string, IList<GitItemStatus>>();
+                    var isMergeCommit = revision.ParentGuids.Count() == 2;
```
Comment (jbialobr) [inline GitUI/UserControls/FileStatusList.cs]: I think this is an unintended change introduced while solving rebase conflicts. The combined diff was moved to the end of the list.

File: GitUI/UserControls/FileStatusList.cs (hunk 2 above)
Comment (vbjay) [inline GitUI/UserControls/FileStatusList.cs]: [EditorConfig extension](https://github.com/editorconfig/editorconfig-visualstudio#readme) can help.

Comment (Nochdarus) [inline GitUI/UserControls/FileStatusList.cs]: Thanks, fixed. Thanks to @vbjay for the tip. In this case I should have just run ctrl+k, ctrl+f, would be enough.

File: GitUI/UserControls/FileStatusList.cs (hunk 3 above)
Comment (Nochdarus) [inline GitUI/UserControls/FileStatusList.cs]: Exactly, it was a mistake during solving rebase conflict. Fixed.

File: GitUI/UserControls/FileStatusList.cs (hunk 1 above)
Comment (Nochdarus) [inline GitUI/UserControls/FileStatusList.cs]: If we assign to null, then we need to explicitly check for it. While new Regex(".*") accepts all files already.
I prefer avoiding null values for logical checks if it is possible.

Comment (Nochdarus) [inline GitUI/UserControls/FileStatusList.cs]: But if it causes the performance issue, then of course I'll change that.

**Comment by Nochdarus** [comment]
I addressed performance issue and review comments. Thank you very much for reviewing.

**Comment by jbialobr** [comment]
>What is the expected behavior with stage all / unstage all in the Commit Window?
 
When I want to stage all files it is unlikely that I perform any filtering before, so
I vote for stage all the files that are visible after aplaying the filter.
It would be nice to show in someway that there are filtred out files.
What about automatically clearing the filter after staging all files?
I am not sure if we need the filter on the staged files, but if yes, it should act in the
same way as for the unstaged files.

**Comment by Nochdarus** [comment]
I think it is fine also to stage all/unstage all regardless of the filter.
And then as you suggest to clear the filters to show that it is actually everything.

And if the user wants to stage only filtered files, then he can do filter + select + stage.

**Comment by jbialobr** [comment]
Yes, and it should clear filter of the source. But I would not enable filtering on staged files.
@pmiossec Any thoughts on this?

**Comment by Nochdarus** [comment]
We can simply hide filtering box for Staged files and that's it.

**Comment by Nochdarus** [comment]
Implemented as discussed: 
Visible only in main window (FormBrowse) and in unstaged files list in Commit Window.
Staging all clears the filter in unstaged.

Default value for other FileStatusList objects is now to hide filter. I mention this because FileStatusList is used in some other places (except for those two mentioned above):

Find all "FileStatusList();", Subfolders, Find Results 1, Entire Solution, ""
**\gitextensions\GitUI\CommandsDialogs\FormCommit.Designer.cs(101):** 
this.Staged = new GitUI.FileStatusList();
**\gitextensions\GitUI\CommandsDialogs\FormDiff.Designer.cs(34):**
this.DiffFiles = new GitUI.FileStatusList();
**\gitextensions\GitUI\CommandsDialogs\FormLog.Designer.cs(36):**
this.DiffFiles = new GitUI.FileStatusList();
**\gitextensions\GitUI\CommandsDialogs\FormStash.Designer.cs(47):**
this.Stashed = new GitUI.FileStatusList();
**\gitextensions\GitUI\CommandsDialogs\RepoHosting\ViewPullRequestsForm.Designer.cs(36):**           this._fileStatusList = new GitUI.FileStatusList();
**\gitextensions\GitUI\HelperDialogs\FormCommitDiff.Designer.cs(34):**
this.DiffFiles = new GitUI.FileStatusList();
  Matching lines: 6    Matching files: 6    Total files searched: 1103

If you think that it makes sense to have filter in any of those, let me know!

**Comment by jbialobr** [comment]
Thanks! I think these two places are enough.

**Pull request merged**